local-cli/generator/templates.js

```diff
@@ -60,50 +63,131 @@ function createProjectFromTemplate(destPath, newProjectName, templateKey, yarnVe
-    // all templates) and every other template only contains additional JS code.
-    // Reason:
-    // This way we don't have to duplicate the native files in every template.
-    // If we duplicated them we'd make RN larger and risk that people would
-    // forget to maintain all the copies so they would go out of sync.
-    const templateName = availableTemplates[templateKey];
-    if (templateName) {
-      copyProjectTemplateAndReplace(
-        path.resolve(
-          'node_modules', 'react-native', 'local-cli', 'templates', templateName
-        ),
-        destPath,
-        newProjectName
-      );
+  if (templateKey === undefined) {
+    // No specific template, use just the HelloWorld template above
+    return;
+  }
+
+  // Keep the files from the 'HelloWorld' template, and overwrite some of them
+  // with the specified project template.
+  // The 'HelloWorld' template contains the native files (these are used by
+  // all templates) and every other template only contains additional JS code.
+  // Reason:
+  // This way we don't have to duplicate the native files in every template.
+  // If we duplicated them we'd make RN larger and risk that people would
+  // forget to maintain all the copies so they would go out of sync.
+  const builtInTemplateName = builtInTemplates[templateKey];
+  if (builtInTemplateName) {
+    // templateKey is e.g. 'navigation',
+    // use the built-in local-cli/templates/HelloNavigation folder
+    createFromBuiltInTemplate(builtInTemplateName, destPath, newProjectName, yarnVersion);
+  } else {
+    // templateKey is e.g. 'ignite',
+    // use the template react-native-template-ignite from npm
+    createFromNpmTemplate(templateKey, destPath, newProjectName, yarnVersion);
+  }
+}
+
+// (We might want to get rid of built-in templates in the future -
+// publish them to npm and install from there.)
+function createFromBuiltInTemplate(templateName, destPath, newProjectName, yarnVersion) {
+  const templatePath = path.resolve(
+    'node_modules', 'react-native', 'local-cli', 'templates', templateName
+  );
+  copyProjectTemplateAndReplace(
+    templatePath,
+    destPath,
+    newProjectName,
+  );
+  installTemplateDependencies(templatePath, yarnVersion);
+}
+
+function createFromNpmTemplate(templateName, destPath, newProjectName, yarnVersion) {
+  const packageName = 'react-native-template-' + templateName;
+  // Check if the template exists
+  console.log(`Fetching template ${packageName} from npm...`);
+  try {
+    if (yarnVersion) {
+      execSync(`yarn info ${packageName}`);
+    } else {
+      execSync(`npm info ${packageName}`);
+    }
+  } catch (err) {
+    throw new Error(`The template ${packageName} was not found on npm:\n` + err.message);
+  }
+  const templatePath = path.resolve(
+    'node_modules', packageName
+  );
+  try {
+    if (yarnVersion) {
+      execSync(`yarn add ${packageName} --ignore-scripts`, {stdio: 'inherit'});
```

Comment:
Thanks, this is a very good point. I'll add support for using templates from local folders.